src/main/java/org/variantsync/functjonal/list/FilterMappedList.java

```diff
@@ -0,0 +1,229 @@
+ * of this class, as this implementation doesn't make any assumption about this.
+ *
+ * @author Benjamin Moosherr
+ */
+public class FilterMappedList<Orig, New> extends AbstractSequentialList<New> {
+    private List<Orig> backingList;
+    private Function<Orig, Optional<New>> filterMap;
+
+    private FilterMappedList(List<Orig> backingList, Function<Orig, Optional<New>> filterMap) {
+        this.backingList = backingList;
+        this.filterMap = filterMap;
+    }
+
+    /**
+     * Creates an unmodifiable view of the elements of {@code list} which satisfy
+     * {@code predicate}.
+     */
+    public static <E> FilterMappedList<E, E> filter(List<E> list, Predicate<E> predicate) {
+        return new FilterMappedList<>(list, (element) -> {
+            if (predicate.test(element)) {
+                return Optional.of(element);
+            } else {
+                return Optional.empty();
+            }
+        });
+    }
+
+    /**
+     * Creates an unmodifiable view of the results of {@code mapper} applied to all elements of
+     * {@code list}
+     */
+    public static <Orig, New> FilterMappedList<Orig, New> map(List<Orig> list, Function<Orig, New> mapper) {
+        return new FilterMappedList<>(list, (element) -> Optional.of(mapper.apply(element)));
+    }
+
+    /**
+     * Creates an unmodifiable view of the present results of {@code mapper} applied to all
+     * elements of {@code list}
+     */
+    public static <Orig, New> FilterMappedList<Orig, New> filterMap(List<Orig> list, Function<Orig, Optional<New>> filterMap) {
+        return new FilterMappedList<>(list, filterMap);
+    }
+
+    private class FilterMapListIterator implements ListIterator<New> {
+        private ListIterator<Orig> backingIterator;
+        private New next;
+        private New previous;
+
+        int nextIndex;
+
+        public FilterMapListIterator(ListIterator<Orig> iterator, int index) {
+            this.backingIterator = iterator;
+            this.nextIndex = index;
+            this.next = null;
+            this.previous = null;
+            this.nextIndex = 0;
+
+            for (int i = 0; i < index; i++) {
+                next();
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (next == null) {
+                next = retrieveNext();
+            }
+
+            return next != null;
+        }
+
+        @Override
+        public boolean hasPrevious() {
+            if (previous == null) {
+                previous = retrievePrevious();
+            }
+
+            return previous != null;
+        }
+
+        @Override
+        public int nextIndex() {
+            return nextIndex;
+        }
+
+        @Override
+        public int previousIndex() {
+            return nextIndex - 1;
+        }
+
+        @Override
+        public New next() {
+            if (next == null) {
```

Comment:
What's the meaning of `next == null`? Please document the idea behind your `next` and `previous` variables on how they cache values returned from the inner iterator.

Reply:
I agree that this might be a bit confusing. In addition to adding documentation, I also refactored this function (and the `previous` function) to be more linear and avoid some duplication.